### DISCUSSION TITLE
Build boot and dtbo

### DISCUSSION THREAD

**Discussion by julianfairfax** [PNDeb/pinenote-debian-image#61] [General]
Your image (which I have cleaned up a lot to use it as a base to port Mobian to the device) generation process generates images for `uboot`, `trust`, `logo` (if you want), and `os1`. In addition, `data` can be wiped. `waveform` shouldn't be touched and you have an image for `uboot_env` too that I haven't used because I don't know what it does.

My question is: is it possible to build `boot` and `dtbo` as well, or is there some reason that wouldn't be possible? I see your `table_standard2` doesn't include these two partitions. Are there needed? Are any of these needed, or could you theoretically use a normal Linux partition table that still includes uboot and the contents of `waveform`, just in other partitions?

**Comment by julianfairfax** [comment]
@m-weigand Another question I have is: is the fact that (at least on the image with [my changes](https://github.com/julianfairfax/pinenote-debian-image/commit/25226300dba9da46dbf27b85c9518d5e8dbedaec)) the device doesn't actually show the OS booting (on the screen) until after running the first boot script (and rebooting) because of the first boot script itself (the systemd unit is supposed to run before everything else), or does the screen not work at all on the OS until `waveform` is "downloaded" for the ebc driver?

**Reply by m-weigand** [reply]
First of all, sorry for not replying earlier, I somehow missed the notification and was just not aware of your questions.

Answering this question first: The problem is that we are not allowed to distribute the waveforms - therefore the Debian image does not come with them and needs to extract the waveform from the waveform partition first, before rebuilding the initrd. Only then can the boot process be shown. See here: https://github.com/PNDeb/pinenote-debian-image/blob/9d80c5aa721e0d70f9ed8f9a943362b44add5420/overlays/root/first_startup.sh#L25

**Reply by m-weigand** [reply]
With regard to your first question:

* dtbo and boot partitions are not required to boot linux, these are artifacts from the original android install
* technically, you also do not need the trust partition any more, as far as I understand the u-boot boot process
* the `uboot_env` partition just holds the u-boot environment and is used to persistently switch the boot order (it's a vfat partition that we can modify from linux. See this script: https://github.com/PNDeb/pinenote-debian-image/blob/trixie/overlays/root/uboot_change_bootmenu.sh

**Reply by m-weigand** [reply]
Maybe just as an additional note: Yesterday I also fixed the generation of a new logo image, which only uses free sources and the Pine64 logo, which is permitted for these kinds. Note that the u-boot distributed with the image will not show logos without the updated (larger, 32mb) partition:
https://github.com/PNDeb/pinenote-debian-image/blob/trixie/.github/workflows/uboot_eink_logo_partition.yml

**Comment by julianfairfax** [comment]
Maybe I did something wrong, but I can't get my PineNote to boot to with [my changes to remove `boot`, `dtbo`, `trust`, and `uboot_env`](https://github.com/julianfairfax/pinenote-debian-image/commit/e4707ee9f391707ac432707ab4c16b73a59ef875). It actually seems the device boots, the first boot script is run, the device reboots, but then I don't see the boot sequence on the screen.

**Reply by julianfairfax** [reply]
I see the problem. `waveform` is now at `/dev/mmcblk0p2`.

**Comment by julianfairfax** [comment]
Thanks for your help. I now have my PineNote up and running with just four partitions: `uboot`, `waveform`, `logo`, and `os`. This should make making a Mobian image in the future (once I have ported the kernel patches over) a lot easier.

I'm going to close this because it's solved. But I do still have a few questions: can `uboot` and `logo` be combined into a single `boot` partition, like how normal Linux partition tables look? And, is the waveform firmware different for each device (like the PineBuds Pro firmware), or the same for all PineNotes (if so, it should be possible to get permission to distribute it, although I still haven't gotten permission to distribute the PineTab2 Wi-Fi driver firmware)?

Lastly, if you look at my image (specifically at the [overlays](https://github.com/julianfairfax/pinenote-debian-image/tree/trixie/overlays)), do you see anything else I could remove? My goal is to port Mobian to PineNote, which means porting the kernel patches, but otherwise as little as possible device-specific changes.

**Reply by m-weigand** [reply]
uboot partition is vfat, logo is a binary partition, that would require modifying the u-boot ebc driver. I will keep it in mind if I ever find time to port the ebc stuff to the upstream u-boot.

**Reply by m-weigand** [reply]
waveforms are, in general, device specific. However, it seems like most pinenotes used the same waveforms. As far as I understand, the waveforms are considered to be intellectual property of eink. However, I'm not a lawyer...

**Reply by m-weigand** [reply]
How do you deal with mesa? I'm not sure the pn will show anything without the mesa-ebc patch

**Reply by julianfairfax** [reply]
> How do you deal with mesa? I'm not sure the pn will show anything without the mesa-ebc patch

Well to be honest I don't know. My image is working for me (although with the 6.3 kernel because I didn't upgrade using your package repo). I don't know why it is working or not working.

**Comment by m-weigand** [comment]
Please note that I uploaded my individual branches, all based upon v6.9, to my kernel repo

https://github.com/m-weigand/linux/branches

The current v6.9 kernel is then generated by this order of merges:
```
git merge branch_rebase_6.9_btpen
git merge branch_rebase_6.9_cyttsp5
git merge branch_rebase_6.9_defconfig
git merge branch_rebase_6.9_ebc
git merge branch_rebase_6.9_iio_accel
git merge branch_rebase_6.9_lm3630a
git merge branch_rebase_6.9_rga2
# rk817 has not pending patches
# git merge branch_rebase_6.9_rk817
git merge branch_rebase_6.9_tps65185
git merge branch_rebase_6.9_quartz64
# only the last two gave merge problems
git merge branch_rebase_6.9_pn_dts_v2
git merge branch_rebase_6.9_rk_suspend_driver
```

**Reply by julianfairfax** [reply]
The image still runs 6.3 though?

**Reply by m-weigand** [reply]
it should use the 6.9 kernel, but have not had time to check

**Reply by m-weigand** [reply]
https://github.com/PNDeb/metapackage-linux-pinenote/blob/debian/latest/debian/control

**Reply by julianfairfax** [reply]
> https://github.com/PNDeb/metapackage-linux-pinenote/blob/debian/latest/debian/control

I see. The image first installs 6.3 though: https://github.com/PNDeb/pinenote-debian-image/blob/trixie/prep_00_get_kernel_files.sh

**Reply by m-weigand** [reply]
that is only for the emergency kernel

**Reply by julianfairfax** [reply]
> that is only for the emergency kernel

Interesting. As you can see above, I don't install the kernel from your repo on my image, so it's using that one then (will obviously change this but I want to port the patches to the Mobian kernel repo and use it anyway).

**Reply by julianfairfax** [reply]
> Please note that I uploaded my individual branches, all based upon v6.9, to my kernel repo
> 
> https://github.com/m-weigand/linux/branches
> 
> The current v6.9 kernel is then generated by this order of merges:
> 
> ```
> git merge branch_rebase_6.9_btpen
> git merge branch_rebase_6.9_cyttsp5
> git merge branch_rebase_6.9_defconfig
> git merge branch_rebase_6.9_ebc
> git merge branch_rebase_6.9_iio_accel
> git merge branch_rebase_6.9_lm3630a
> git merge branch_rebase_6.9_rga2
> # rk817 has not pending patches
> # git merge branch_rebase_6.9_rk817
> git merge branch_rebase_6.9_tps65185
> git merge branch_rebase_6.9_quartz64
> # only the last two gave merge problems
> git merge branch_rebase_6.9_pn_dts_v2
> git merge branch_rebase_6.9_rk_suspend_driver
> ```

I can't even clone just the repo itself:

```
$ git clone https://github.com/m-weigand/linux.git
Klone nach 'linux'...
remote: Enumerating objects: 10197515, done.
Fehler: RPC fehlgeschlagen; curl 92 HTTP/2 stream 5 was not closed cleanly: CANCEL (err 8)
Fehler: 40 Bytes des Bodys werden noch erwartet
fetch-pack: unexpected disconnect while reading sideband packet
Schwerwiegend: zu frühes Dateiende
Schwerwiegend: fetch-pack: ungültige index-pack Ausgabe
```

I appreciate that this branch system is your preferred method of working with all the patches the PineNote needs, but is it all possible to condense this down to just one branch, or even better, one commit? I would also like to use the least amount of patches possible, since adding each one to Mobian's kernel repo is already not easy.

**Reply by m-weigand** [reply]
`branch_pinenote_6-9_v1` is the branch with all the relevant branches merged into. This is the back that I use to generate debian packages for the PN from.

The clone issue seems to be an issue of slow internet connections... I was able to solve this by falling back to http v1, as suggested by the internet:
```
git config --global http.version HTTP/1.1
git clone --depth 1 -b branch_pinenote_6-9_v1 https://github.com/m-weigand/linux.git
cd linux
git fetch --unshallow
git config --global http.version HTTP/2
```

**Reply by julianfairfax** [reply]
Thanks for the tips! I was able to clone the repo and turn all of your commits into [a single one](https://github.com/julianfairfax/linux/commit/a4e82a6d0c29d73f31c0a762a93b89919c8d4f92). I must say though, the number of drivers is a lot. I guess none of these were ever upstreamed. `include` contains driver headers, right? I believe we can do without the documentation (not that's a big issue to add)?

I will try to add these files to Mobian's rockchip kernel. Hopefully the changes don't cause any issues for other devices.

**Reply by julianfairfax** [reply]
Added those changes to Mobian's rockchip, using [this patch file](https://github.com/user-attachments/files/16418866/0001-Add-support-for-the-PineNote.patch.txt). Unfortunately it doesn't boot. [Here](https://github.com/user-attachments/files/16418893/linux-6.9.0-rc4-gcfa5ba286863-dirty_6.9.0-rc4-gcfa5ba286863-5_arm64.zip). I built using the `pinenote_defconfig`, so a bit surprised it doesn't boot. I would also need to know what exactly is needed from the defconfig, so I can add those changes to Mobian's `rockchip_defconfig` (so we can build one rockchip kernel for all devices).

Thanks for your help with all this!

**Reply by m-weigand** [reply]
what does the uart say when the boot fails?

**Reply by julianfairfax** [reply]
> what does the uart say when the boot fails?

That's the thing. I got this device second-hand and don't have uart adaptor (have one from my PineTab2, which probably won't work?). Does that make this whole porting a big waste of time?

**Reply by talpadk** [reply]
If you are not afraid of a bit of soldering you can relatively easily  make your own UART adapter.

https://wiki.pine64.org/wiki/PineNote_Development/UART

It also has the added benefit that you can have a USB data connection at the same time as the UART connection.
It is actually my preferred way to get an UART connection to my PN.
(You must use a USB-A -> USB-C cable for the data connection as the interface uses the CC1, CC2, SBU1 and SBU2 on the pines USB-C)

Admittedly I'm considering making a proper PCB that integrates it all but for now my dangling wire interface is fine

**Reply by julianfairfax** [reply]
Now experiencing another issue while trying to build a trixie image from my fork with your latest kernel: https://github.com/julianfairfax/pinenote-debian-image/actions/runs/10291089615/job/28482581642.

It seems also that, without the UART adaptor, porting the kernel patches to the Mobian kernel is not practical. I also don't particularly want to maintain them.

My PineNote was actually given to me, but I may give it to someone else who does have time and willingness to work on this more (the this being official Mobian support).

**Reply by m-weigand** [reply]
Maybe you want to contact me via matrix?

The image is currently failing, yes. That is because I did not yet build new gnome/mesa packages yet (due to summer vacation).

**Reply by julianfairfax** [reply]
> Maybe you want to contact me via matrix?
> 
What is your Matrix handle?

**Reply by m-weigand** [reply]
@mweigand:matrix.org
